client/src/controllers/LoaderController.ts

```diff
@@ -0,0 +1,121 @@
+    durationSeconds: { type: Number, default: 30 },
+  };
+
+  declare activeClass: string;
+  declare activeValue: string;
+  declare durationSecondsValue: number;
+  declare labelTarget: HTMLElement;
+  declare hasActiveClass: boolean;
+  declare hasLabelTarget: boolean;
+  declare originalText: string;
+
+  static afterLoad(identifier: string, application: Application) {
+    const { controllerAttribute } = application.schema;
+    const { actionAttribute } = application.schema;
+
+    const updateLoader = () => {
+      const em = document.querySelector('em') as HTMLElement;
+      document
+        .querySelectorAll(
+          `.button-longrunning:not([data-controller~='${identifier}'])`,
+        )
+        .forEach((button) => {
+          button.setAttribute(controllerAttribute, identifier);
+          button.setAttribute(actionAttribute, 'w-loader#activate');
+
+          if (button.getAttribute('data-clicked-text')) {
+            button.setAttribute(
+              'data-w-loader-active-value',
+              `${button.getAttribute('data-clicked-text')}`,
+            );
+            button.removeAttribute('data-clicked-text');
+          }
+
+          if (em.matches('button em')) {
+            em.setAttribute('data-w-loader-target', 'label');
+          }
+
+          button.setAttribute('data-w-loader-duration-seconds-value', '10');
+        });
+    };
+
+    document.addEventListener('DOMContentLoaded', updateLoader);
+  }
+
+  activate() {
+    this.originalText = this.hasLabelTarget
+      ? this.labelTarget.innerText
+      : (this.element.textContent as string);
+
+    // If client-side validation is active on this form, and is going to block submission of the
+    // form, don't activate the spinner
+    const form = this.element.closest('form');
+
+    if (
+      form &&
+      form.checkValidity &&
+      !form.noValidate &&
+      !form.checkValidity()
+    ) {
+      return;
+    }
+
+    // set time out
+    this.setTimeout();
+  }
+
+  cancelSpinner(originalText: string) {
+    this.element.removeAttribute('disabled');
+    this.element.classList.remove(
+      this.hasActiveClass ? this.activeClass : 'button-longrunning-active',
+    );
+
+    if (this.activeValue && this.hasLabelTarget) {
+      this.labelTarget.innerText = originalText;
+    }
+  }
+
+  setTimeout() {
+    const timeout = setTimeout(() => {
+      if (this.activeValue && this.hasLabelTarget === true) {
+        // Sets the targetElement innerText to the activeValue
+        this.labelTarget.innerText = this.activeValue;
+      }
+
+      // Disabling button must be done last: disabled buttons can't be
+      // modified in the normal way, it would seem.
+      this.element.classList.add(
+        this.hasActiveClass ? this.activeClass : 'button-longrunning-active',
+      );
+      this.element.setAttribute('disabled', '');
+
+      setTimeout(() => {
+        this.cancelSpinner(this.originalText);
```

Comment:
When you are passing a param that is just from `this` - it's usually much simpler to grab the value in the method itself.